File: e2e/src/ez_cdc_e2e/datasources/store.py
"""Persistence layer for the datasources YAML file.

DatasourceStore wraps a single `e2e/ez-cdc.yaml` file and provides:

  - load() / reload() — read from disk and validate
  - save() — atomic write back to disk (write to temp + rename)
  - add_source(name, spec) / add_sink(name, spec) — add or replace
  - remove(name) — drop a source or sink by name
  - get_source(name) / get_sink(name) — fetch with clear errors
  - list_sources() / list_sinks() — sorted name lists
  - has_any() — true if at least 1 source AND 1 sink
  - exists(name) — true if any datasource (source or sink) has that name

Atomicity:
  Writes go through a temp file in the same directory followed by os.rename
  to avoid leaving a half-written YAML on crash. A `.bak` snapshot of the
  previous content is kept next to the file (overwritten on each save) so
  the user can recover from a wizard mistake.

Threading:
  Not thread-safe. The CLI is single-threaded; if that ever changes, wrap
  the store in a threading.Lock at the call site.
"""

from __future__ import annotations

import os
import shutil
from pathlib import Path
from typing import Optional

import yaml

from .loader import (
    DatasourceError,
    DatasourceNotFoundError,
    load_datasources,
)
from .schema import (
    DatasourcesFile,
    PostgresSinkSpec,
    PostgresSourceSpec,
    SinkSpec,
    SnowflakeSinkSpec,
    SourceSpec,
    StarRocksSinkSpec,
)


class DatasourceStore:
    """File-backed CRUD for datasources.

    Constructed with a path. The file may not exist yet — use `load()` with
    `allow_missing=True` to handle the "first run" case gracefully, or
    construct an empty store with `DatasourceStore.empty(path)`.
    """

    def __init__(self, path: Path, env_files: Optional[list[Path]] = None) -> None:
        self.path = Path(path).resolve()
        self.env_files = env_files or []
        self._data: DatasourcesFile = DatasourcesFile()
        self._loaded = False

    # ── factory ──────────────────────────────────────────────────────────────

    @classmethod
    def empty(cls, path: Path) -> "DatasourceStore":
        """Build an in-memory empty store. Doesn't read disk. Useful for the
        first-run flow where the wizard creates datasources from scratch."""
        store = cls(path)
        store._data = DatasourcesFile()
        store._loaded = True
        return store

    # ── loading ──────────────────────────────────────────────────────────────

    def load(self, *, allow_missing: bool = True) -> None:
        """Read and validate the file. Replaces in-memory state.

        With allow_missing=True (default), a non-existent file becomes an
        empty DatasourcesFile rather than raising. This is the right
        behavior for the gate that asks "do we have datasources yet?"
        """
        self._data = load_datasources(
            self.path,
            env_files=self.env_files,
            allow_missing=allow_missing,
        )
        self._loaded = True

    def reload(self) -> None:
        """Re-read the file from disk, discarding any in-memory mutations."""
        self.load(allow_missing=True)

    @property
    def loaded(self) -> bool:
        return self._loaded

    @property
    def data(self) -> DatasourcesFile:
        """Direct access to the underlying DatasourcesFile.

        Use this for read-only operations that need fields not exposed by
        the convenience methods. For mutations, prefer the `add_*` / `remove`
        methods so the change goes through the validator and gets persisted.
        """
        if not self._loaded:
            self.load()
        return self._data

    # ── inspection ───────────────────────────────────────────────────────────

    def has_any(self) -> bool:
        return self.data.has_any()

    def is_empty(self) -> bool:
        return self.data.is_empty()

    def list_sources(self) -> list[str]:
        return self.data.list_source_names()

    def list_sinks(self) -> list[str]:
        return self.data.list_sink_names()

    def get_source(self, name: str) -> SourceSpec:
        return self.data.get_source(name)

    def get_sink(self, name: str) -> SinkSpec:
        return self.data.get_sink(name)

    def exists(self, name: str) -> bool:
        """True if any datasource (source or sink) has this name. Used by the
        wizard to detect collisions before saving."""
        d = self.data
        return name in d.sources or name in d.sinks

    # ── mutation ─────────────────────────────────────────────────────────────

    def add_source(self, name: str, spec: SourceSpec, *, replace: bool = False) -> None:
        """Add a source datasource. Raises if the name already exists in
        either sources or sinks (use replace=True to override an existing source)."""
        self._check_name_collision(name, kind="source", replace=replace)
        self.data.sources[name] = spec  # pydantic re-validates because validate_assignment=True

    def add_sink(self, name: str, spec: SinkSpec, *, replace: bool = False) -> None:
        """Add a sink datasource. Same collision rules as add_source."""
        self._check_name_collision(name, kind="sink", replace=replace)
        self.data.sinks[name] = spec

    def remove(self, name: str) -> None:
        """Remove a datasource by name (source or sink). Raises if not found."""
        if name in self.data.sources:
            del self.data.sources[name]
            return
        if name in self.data.sinks:
            del self.data.sinks[name]
            return
        raise DatasourceNotFoundError(
            f"datasource {name!r} not found in {self.path}"
        )

    def _check_name_collision(self, name: str, *, kind: str, replace: bool) -> None:
        d = self.data
        # Cross-kind collision: a name can't be both a source and a sink.
        if kind == "source" and name in d.sinks:
            raise DatasourceError(
                f"name {name!r} is already used by a sink. "
                f"Choose a different name or remove the sink first."
            )
        if kind == "sink" and name in d.sources:
            raise DatasourceError(
                f"name {name!r} is already used by a source. "
                f"Choose a different name or remove the source first."
            )
        # Same-kind collision: only allowed with replace=True.
        existing = d.sources if kind == "source" else d.sinks
        if name in existing and not replace:
            raise DatasourceError(
                f"{kind} {name!r} already exists. Pass replace=True to overwrite, "
                f"or remove it first."
            )

    # ── persistence ──────────────────────────────────────────────────────────

    def save(self, *, mode: int = 0o600) -> None:
        """Write the in-memory state back to disk atomically.

        Steps:
          1. Serialize to YAML.
          2. Write to a temporary file in the same directory.
          3. fsync the temp file.
          4. Backup the existing file to .bak (if any).
          5. os.rename(temp, target) — atomic on POSIX.
          6. chmod the target to `mode` (default 0o600 — credentials).

        On any failure, the original file is left untouched.
        """
        self.path.parent.mkdir(parents=True, exist_ok=True)
        yaml_text = self._render_yaml()

        tmp_path = self.path.with_suffix(self.path.suffix + ".tmp")
        try:
            with open(tmp_path, "w", encoding="utf-8") as f:
                f.write(yaml_text)
                f.flush()
                try:
                    os.fsync(f.fileno())
                except OSError:
                    pass  # not all filesystems support fsync; non-fatal

            # Backup the existing file before overwriting.
            if self.path.exists():
                bak_path = self.path.with_suffix(self.path.suffix + ".bak")
                try:
                    shutil.copy2(self.path, bak_path)
                except OSError:
                    pass  # backup failure is non-fatal — proceed with the save

            os.replace(tmp_path, self.path)

            try:
                os.chmod(self.path, mode)
            except OSError:
                pass  # non-POSIX filesystem; non-fatal

        finally:
            # Clean up the temp file if it's still there (e.g., on error).
            if tmp_path.exists():
                try:
                    tmp_path.unlink()
                except OSError:
                    pass

    def _render_yaml(self) -> str:
        """Convert the in-memory DatasourcesFile to a YAML string.

        We render via pydantic's model_dump (with by_alias=True so the
        `schema_` field becomes `schema:` in the output) and then call
        yaml.safe_dump with sort_keys=False to preserve our intended order.
        """
        # by_alias=True so PostgresSinkSpec.schema_ → "schema" in output.
        # exclude_none=True so optional fields aren't littered as `null`.
        as_dict = self._data.model_dump(by_alias=True, exclude_none=True)

        # Render with a header comment for context.
        header = (
            "# ez-cdc configuration file\n"
            "# Datasources + pipeline settings. Managed by `ez-cdc datasource ...`\n"
            "# or edited manually. See e2e/ez-cdc.example.yaml for the schema.\n"
            "# Credentials may use ${VAR} interpolation; see ez-cdc docs.\n"
            "\n"
        )

        body = yaml.safe_dump(
            as_dict,
            sort_keys=False,
            default_flow_style=False,
            allow_unicode=True,
            indent=2,
        )

        return header + body

File: e2e-cli/src/config/store.rs
//! Persistence layer for the datasources YAML file.
//!
//! `DatasourceStore` wraps `ez-cdc.yaml` and provides load/save/CRUD
//! with atomic writes (temp file + rename) and `.bak` backups.

use std::path::{Path, PathBuf};

use super::loader::{load_datasources, DatasourceError};
use super::schema::{DatasourcesFile, SinkSpec, SourceSpec};

pub struct DatasourceStore {
    path: PathBuf,
    env_files: Vec<PathBuf>,
    data: DatasourcesFile,
    loaded: bool,
}

impl DatasourceStore {
    pub fn new(path: &Path, env_files: Vec<PathBuf>) -> Self {
        Self {
            path: path.to_path_buf(),
            env_files,
            data: DatasourcesFile::default(),
            loaded: false,
        }
    }

    /// Build an in-memory empty store without reading disk.
    pub fn empty(path: &Path) -> Self {
        Self {
            path: path.to_path_buf(),
            env_files: vec![],
            data: DatasourcesFile::default(),
            loaded: true,
        }
    }

    pub fn path(&self) -> &Path {
        &self.path
    }

    // ── loading ────────────────────────────────────────────────────────────

    /// Read and validate the file. Replaces in-memory state.
    pub fn load(&mut self, allow_missing: bool) -> Result<(), DatasourceError> {
        let env_refs: Vec<&Path> = self.env_files.iter().map(|p| p.as_path()).collect();
        self.data = load_datasources(&self.path, &env_refs, allow_missing)?;
        self.loaded = true;
        Ok(())
    }

    /// Re-read from disk, discarding in-memory mutations.
    pub fn reload(&mut self) -> Result<(), DatasourceError> {
        self.load(true)
    }

    /// Ensure the store is loaded (lazy load on first access).
    fn ensure_loaded(&mut self) -> Result<(), DatasourceError> {
        if !self.loaded {
            self.load(true)?;
        }
        Ok(())
    }

    pub fn data(&mut self) -> Result<&DatasourcesFile, DatasourceError> {
        self.ensure_loaded()?;
        Ok(&self.data)
    }

    // ── inspection ─────────────────────────────────────────────────────────

    pub fn has_any(&mut self) -> Result<bool, DatasourceError> {
        Ok(self.data()?.has_any())
    }

    pub fn is_empty(&mut self) -> Result<bool, DatasourceError> {
        Ok(self.data()?.is_empty())
    }

    pub fn list_sources(&mut self) -> Result<Vec<String>, DatasourceError> {
        Ok(self.data()?.list_source_names())
    }

    pub fn list_sinks(&mut self) -> Result<Vec<String>, DatasourceError> {
        Ok(self.data()?.list_sink_names())
    }

    pub fn get_source(&mut self, name: &str) -> Result<SourceSpec, DatasourceError> {
        let spec = self.data()?.get_source(name)
            .map_err(|e| DatasourceError::NotFound(e))?;
        Ok(spec.clone())
    }

    pub fn get_sink(&mut self, name: &str) -> Result<SinkSpec, DatasourceError> {
        let spec = self.data()?.get_sink(name)
            .map_err(|e| DatasourceError::NotFound(e))?;
        Ok(spec.clone())
    }

    pub fn exists(&mut self, name: &str) -> Result<bool, DatasourceError> {
        self.ensure_loaded()?;
        Ok(self.data.sources.contains_key(name) || self.data.sinks.contains_key(name))
    }

    // ── mutation ───────────────────────────────────────────────────────────

    pub fn add_source(
        &mut self,
        name: String,
        spec: SourceSpec,
        replace: bool,
    ) -> Result<(), DatasourceError> {
        self.ensure_loaded()?;
        self.check_name_collision(&name, "source", replace)?;
        self.data.sources.insert(name, spec);
        Ok(())
    }

    pub fn add_sink(
        &mut self,
        name: String,
        spec: SinkSpec,
        replace: bool,
    ) -> Result<(), DatasourceError> {
        self.ensure_loaded()?;
        self.check_name_collision(&name, "sink", replace)?;
        self.data.sinks.insert(name, spec);
        Ok(())
    }

    pub fn remove(&mut self, name: &str) -> Result<(), DatasourceError> {
        self.ensure_loaded()?;
        if self.data.sources.remove(name).is_some() {
            return Ok(());
        }
        if self.data.sinks.remove(name).is_some() {
            return Ok(());
        }
        Err(DatasourceError::NotFound(format!(
            "datasource {name:?} not found in {}",
            self.path.display()
        )))
    }

    fn check_name_collision(
        &self,
        name: &str,
        kind: &str,
        replace: bool,
    ) -> Result<(), DatasourceError> {
        // Cross-kind collision.
        if kind == "source" && self.data.sinks.contains_key(name) {
            return Err(DatasourceError::Validation(format!(
                "name {name:?} is already used by a sink. \
                 Choose a different name or remove the sink first."
            )));
        }
        if kind == "sink" && self.data.sources.contains_key(name) {
            return Err(DatasourceError::Validation(format!(
                "name {name:?} is already used by a source. \
                 Choose a different name or remove the source first."
            )));
        }
        // Same-kind collision.
        let existing = if kind == "source" {
            &self.data.sources
        } else {
            // SinkSpec and SourceSpec use different maps but we need a common check.
            // For sinks, check the sinks map. For sources, already handled above.
            // This is a simplified approach — just check the right map.
            return if !replace && self.data.sinks.contains_key(name) {
                Err(DatasourceError::Validation(format!(
                    "{kind} {name:?} already exists. Use replace=true to overwrite."
                )))
            } else {
                Ok(())
            };
        };
        if !replace && existing.contains_key(name) {
            return Err(DatasourceError::Validation(format!(
                "{kind} {name:?} already exists. Use replace=true to overwrite."
            )));
        }
        Ok(())
    }

    // ── persistence ────────────────────────────────────────────────────────

    /// Write in-memory state to disk atomically.
    ///
    /// Steps: write temp → fsync → backup .bak → rename → chmod 0o600.
    pub fn save(&self) -> Result<(), DatasourceError> {
        if let Some(parent) = self.path.parent() {
            std::fs::create_dir_all(parent)
                .map_err(|e| DatasourceError::Io(format!("create dir: {e}")))?;
        }

        let yaml_text = self.render_yaml()?;
        let tmp_path = self.path.with_extension("yaml.tmp");

        // Write to temp file.
        std::fs::write(&tmp_path, &yaml_text)
            .map_err(|e| DatasourceError::Io(format!("write temp file: {e}")))?;

        // Backup existing file.
        if self.path.exists() {
            let bak_path = self.path.with_extension("yaml.bak");
            let _ = std::fs::copy(&self.path, &bak_path); // non-fatal
        }

        // Atomic rename.
        std::fs::rename(&tmp_path, &self.path)
            .map_err(|e| {
                let _ = std::fs::remove_file(&tmp_path);
                DatasourceError::Io(format!("rename: {e}"))
            })?;

        // Set permissions (non-fatal on non-POSIX).
        #[cfg(unix)]
        {
            use std::os::unix::fs::PermissionsExt;
            let _ = std::fs::set_permissions(
                &self.path,
                std::fs::Permissions::from_mode(0o600),
            );
        }

        Ok(())
    }

    fn render_yaml(&self) -> Result<String, DatasourceError> {
        let header = "\
# ez-cdc configuration file\n\
# Datasources + pipeline settings. Managed by `ez-cdc datasource ...`\n\
# Credentials may use ${VAR} interpolation.\n\
# See e2e/README.md for settings documentation.\n\
\n";

        let body = serde_yml::to_string(&self.data)
            .map_err(|e| DatasourceError::Yaml(format!("serialize: {e}")))?;

        Ok(format!("{header}{body}"))
    }
}

// ── Tests ──────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;
    use crate::config::schema::*;

    fn demo_source() -> SourceSpec {
        SourceSpec::Postgres(PostgresSourceInner {
            url: "postgres://localhost/test".into(),
            seed: None,
            replication_slot: "test_slot".into(),
            publication: "test_pub".into(),
            tables: vec!["t1".into()],
        })
    }

    fn demo_sink() -> SinkSpec {
        SinkSpec::Starrocks(StarRocksSinkInner {
            url: "http://localhost:8030".into(),
            mysql_port: 9030,
            database: "test".into(),
            user: "root".into(),
            password: "".into(),
        })
    }

    #[test]
    fn empty_store() {
        let store = DatasourceStore::empty(Path::new("/tmp/test.yaml"));
        assert!(store.data.is_empty());
    }

    #[test]
    fn add_and_remove() {
        let mut store = DatasourceStore::empty(Path::new("/tmp/test.yaml"));
        store.add_source("src1".into(), demo_source(), false).unwrap();
        store.add_sink("sink1".into(), demo_sink(), false).unwrap();
        assert!(store.exists("src1").unwrap());
        assert!(store.exists("sink1").unwrap());
        assert!(!store.exists("nope").unwrap());

        store.remove("src1").unwrap();
        assert!(!store.exists("src1").unwrap());

        assert!(store.remove("nonexistent").is_err());
    }

    #[test]
    fn cross_kind_collision() {
        let mut store = DatasourceStore::empty(Path::new("/tmp/test.yaml"));
        store.add_source("shared".into(), demo_source(), false).unwrap();
        let err = store.add_sink("shared".into(), demo_sink(), false).unwrap_err();
        assert!(err.to_string().contains("already used by a source"));
    }

    #[test]
    fn same_kind_collision() {
        let mut store = DatasourceStore::empty(Path::new("/tmp/test.yaml"));
        store.add_source("dup".into(), demo_source(), false).unwrap();
        assert!(store.add_source("dup".into(), demo_source(), false).is_err());
        // With replace=true, it works.
        store.add_source("dup".into(), demo_source(), true).unwrap();
    }

    #[test]
    fn save_and_reload() {
        let dir = std::env::temp_dir().join("ez-cdc-store-test");
        std::fs::create_dir_all(&dir).unwrap();
        let path = dir.join("ez-cdc.yaml");

        let mut store = DatasourceStore::empty(&path);
        store.add_source("s1".into(), demo_source(), false).unwrap();
        store.add_sink("k1".into(), demo_sink(), false).unwrap();
        store.save().unwrap();

        // Reload from disk.
        let mut store2 = DatasourceStore::new(&path, vec![]);
        store2.load(false).unwrap();
        assert!(store2.exists("s1").unwrap());
        assert!(store2.exists("k1").unwrap());

        let _ = std::fs::remove_dir_all(&dir);
    }

    #[test]
    fn save_creates_backup() {
        let dir = std::env::temp_dir().join("ez-cdc-bak-test");
        std::fs::create_dir_all(&dir).unwrap();
        let path = dir.join("ez-cdc.yaml");

        // First save.
        let store = DatasourceStore::empty(&path);
        store.save().unwrap();
        assert!(path.exists());

        // Second save — should create .bak.
        store.save().unwrap();
        let bak = dir.join("ez-cdc.yaml.bak");
        assert!(bak.exists());

        let _ = std::fs::remove_dir_all(&dir);
    }
}
